fe/fe-core/src/main/java/com/starrocks/qe/QueryQueueManager.java

```diff
@@ -77,140 +49,78 @@ public static QueryQueueManager getInstance() {
-
-            node.updateResourceUsage(numRunningQueries, memLimitBytes, memUsedBytes, cpuUsedPermille);
-            maybeNotifyAfterLock();
-        } finally {
-            lock.unlock();
-        }
-    }
-
-    public void maybeWait(ConnectContext connectCtx, DefaultCoordinator coord) throws UserException, InterruptedException {
-        if (!needCheckQueue(coord)) {
-            return;
-        }
-        if (!enableCheckQueue(coord) || canRunMore()) {
+    public void maybeWait(ConnectContext context, DefaultCoordinator coord) throws UserException, InterruptedException {
+        if (!needCheckQueue(coord) || !isEnableQueue(coord)) {
             return;
         }
 
         long startMs = System.currentTimeMillis();
-        long timeoutMs;
-        PendingQueryInfo info = new PendingQueryInfo(connectCtx, lock, coord);
         boolean isPending = false;
-
         try {
-            lock.lock();
-            if (!enableCheckQueue(coord) || canRunMore()) {
-                return;
-            }
-
-            if (!canQueueMore()) {
-                throw new UserException("Need be queued but exceed query queue capacity");
-            }
+            Slot slotRequirement = createSlot(coord);
+            coord.setSlot(slotRequirement);
+            Future<Slot> slotFuture = GlobalStateMgr.getCurrentState().getSlotProvider().requireSlot(slotRequirement);
 
             isPending = true;
-            info.connectCtx.setPending(true);
-            pendingQueryInfoMap.put(info.connectCtx, info);
+            context.setPending(true);
             MetricRepo.COUNTER_QUERY_QUEUE_PENDING.increase(1L);
             MetricRepo.COUNTER_QUERY_QUEUE_TOTAL.increase(1L);
 
-            while (enableCheckQueue(coord) && !canRunMore()) {
-                timeoutMs = startMs + GlobalVariable.getQueryQueuePendingTimeoutSecond() * 1000L;
+            long timeoutMs = slotRequirement.getExpiredPendingTimeMs();
+            Slot allocatedSlot = null;
+            while (allocatedSlot == null) {
+                // Check timeout.
                 long currentMs = System.currentTimeMillis();
                 if (currentMs >= timeoutMs) {
                     MetricRepo.COUNTER_QUERY_QUEUE_TIMEOUT.increase(1L);
-                    throw new UserException("Pending timeout");
+                    GlobalStateMgr.getCurrentState().getSlotProvider().cancelSlotRequirement(slotRequirement);
+                    String errMsg = String.format(PENDING_TIMEOUT_ERROR_MSG_FORMAT,
+                            GlobalVariable.getQueryQueuePendingTimeoutSecond(),
+                            GlobalVariable.QUERY_QUEUE_PENDING_TIMEOUT_SECOND);
+                    throw new UserException(errMsg);
                 }
 
-                info.await(Math.min(timeoutMs - currentMs, CHECK_INTERVAL_MS), TimeUnit.MILLISECONDS);
+                // Check leader changed.
+                Pair<String, Integer> masterIpAndPort = GlobalStateMgr.getCurrentState().getLeaderIpAndRpcPort();
```

Comment:
I think watch the leader change event is better.

Reply:
Fine. Updated.